src/sgidClient.ts

```
@@ -0,0 +1,233 @@
+   * Defaults to 'myinfo.nric_number openid'.
+   * @param nonce Unique nonce for this request. If this param is undefined, a nonce is generated
+   * and returned. To prevent this behaviour, specify null for this param.
+   * @param redirectUri The redirect URI used in the authorization request. If this param is provided,
+   * it will be used over the redirect URI provided in the SgidClient constructor. The redirect URI
+   * must be provided in at least the SgidClient constructor or this function.
+   * @param codeChallenge The code challenge from the code verifier used for PKCE enhancement
+   */
+  authorizationUrl({
+    state,
+    scope = DEFAULT_SCOPE,
+    nonce = generators.nonce(),
+    redirectUri = this.getFirstRedirectUri(),
+    codeChallenge,
+  }: AuthorizationUrlParams): AuthorizationUrlReturn {
+    const url = this.sgID.authorizationUrl({
+      scope: typeof scope === 'string' ? scope : scope.join(' '),
+      nonce: nonce ?? undefined,
+      state,
+      redirect_uri: redirectUri,
+      code_challenge: codeChallenge,
+      code_challenge_method: DEFAULT_SGID_CODE_CHALLENGE_METHOD,
+    })
+
+    const result: { url: string; nonce?: string } = { url }
+    if (nonce) {
+      result.nonce = nonce
+    }
+    return result
+  }
+
+  private getFirstRedirectUri(): string {
+    if (
+      !this.sgID.metadata.redirect_uris ||
+      this.sgID.metadata.redirect_uris.length === 0
+    ) {
+      throw new Error(Errors.MISSING_REDIRECT_URI_ERROR)
+    }
+    return this.sgID.metadata.redirect_uris[0]
+  }
+
+  /**
+   * Exchanges authorization code for access token.
+   * @param code The authorization code received from the authorization server
+   * @param nonce Nonce passed to authorizationUrl for this request. Specify null
+   * if no nonce was passed to authorizationUrl.
+   * @param redirectUri The redirect URI used in the authorization request. Defaults to the one
```

if redirect URI not provided in constructor or here, do we need to throw an error here too, just like in authorizationUrl?

---

yup i think we need to throw an error because the token endpoint expects a redirect URI. the function `getFirstRedirectUri` is used in both `authorizationUrl` and `callback` and is the function that will throw the error.

---

ah ok thanks for explaining!